// planorama/mongohandler/index.js

// Dependencies

var MongoClient = require('mongodb').MongoClient;
var debug = require('debug')('my-application');
var async = require('async');

/**
 *
 * @param collectionName
 * @param dbName
 * @returns {{}}
 * @constructor
 */

var MongoHandler = function(dbName){


  var handler = {}

  handler.dbName = dbName || 'test';

  handler.getDbPath = function getDbPath(){
    return 'mongodb://127.0.0.1:27017/' + handler.dbName;
  };

  handler.open = function open(callback){
    return MongoClient.connect(handler.getDbPath(), function(err, db){
      if(err) throw err;
      if(!db) throw new Error('Could not get a database', handler.getDbPath());
      handler.db = db;
      console.log('Connected to database: ', handler.getDbPath());
      return callback(null, handler);
    });
  };

  handler.collection = function collection(collectionName){
    if(!handler.db) throw new Error('Database has to be opened first');
    return handler.db.collection(collectionName)
  };

  handler.initCollec = function (collectionName, index_init, callback) {

    var collection = handler.collection(collectionName || 'test');

    async.series([
        function remove_data(cb) {
          return collection.remove(cb);
        },
        function remove_index(cb) {
          return collection.dropAllIndexes(cb);
        },
        function set_index(cb) {
          return index_init(collection, callback);
          return collection.ensureIndex({ "username": 1 }, { unique: true }, cb);
        }
      ],
      callback
    );
  };

  return handler;
}


module.exports = MongoHandler
