// planorama/basicModel/index.js

// Dependencies

var ObjectID = require('mongodb').ObjectID;
var BBO = require('bluebirds').Object;

// Internal requires


/***********************************************************************************************************************
 * BaseModel
 ***********************************************************************************************************************
 *
 * Generic implementation of an ORM-like base model.
 *
 * To create a new model inheriting from BaseModel:
 *
 * var Foo = BaseModel.extend({
 *   // add properties and methods here
 *   bar: 'baz',
 *   speak: function() {
 *     console.log(this.bar);
 *   }
 * },
 * {
 *
 *   _collectionName: 'foos'   // the collection name in MongoDB
 *
 *   // add static properties and methods here
 *
 * })
 *
 **********************************************************************************************************************/
var BaseModel = BBO.extend({

    /**
     * save
     *
     * @info saves into the users collection
     *
     * @param callback
     * @returns {*|Session}
     */
    save: function(callback) {
      return this.constructor.getCollection().save(this, callback);
    }
  },

  /**
  static attributes:
  **/

  {

    _collectionName: null,
    _db: null,

    /**
     * create
     *
     * @info initialization comportment setting for different formats of input parameters
     * @param def
     */
    create: function(def){
      // convert _id as a Mongo ID if provided
      if (def._id) {
        def._id = new ObjectID(def._id);  // works also if def._id is already an ObjectID
      }
      return this._super(def);

    },

    getCollection: function() {
      if (!this._collectionName) throw new Error('Undefined collection name');
      return this._db.collection(this._collectionName);
    },

    /**
     * findOne
     *
     * @param filter
     * @param callback
     * @returns {*}
     */
    findOne: function(filter, callback) {
      var self = this;
      return this.getCollection().findOne(filter, function(err, item){
        err = err || (!item && new Error('No user found'));
        if (err) return callback(err);
        return callback(err, self.create(item));
      });
    },

    /**
     * findById
     *
     * @param id
     * @param callback
     * @returns {*}
     */
    findById: function(id, callback) {
      // Important: make the conversion into a valid Mongo ID before searching
      return this.findOne({_id: new ObjectID(id)}, callback);
    },

    /**
     * count
     *
     * @param query
     * @param callback
     * @returns {*|null}
     */
    count: function(query, callback) {
      return this.getCollection().count(query, callback);
    },

    /**
     * remove
     *
     * @param query
     * @param callback
     * @returns {*}
     */
    remove: function(query, callback) {
      return this.getCollection().remove(query, callback);
    }

  }
);


module.exports = BaseModel;
